File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    let ans = [-1, -1];
    let [l, r] = [0, nums.length - 1];
    let pivot = Math.floor((r + l) / 2);
    while (l <= r) {
        console.log(l, r, pivot, nums[pivot]);
        if (nums[pivot] !== target) {
            if (nums[pivot] > target) {
                r = pivot - 1;
            } else if (nums[pivot] < target) {
                l = pivot + 1;
            }
            pivot = Math.floor((r + l) / 2);
        } else {
            [l, r] = [pivot, pivot];
            while (nums[l] === target) {
                l--;
            }
            while (nums[r] === target) {
                r++;
            }
            ans = [l + 1, r - 1];
            break;
        }
    }
    return ans;
};

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.go
func searchRange(nums []int, target int) []int {
  l, r := 0, len(nums)-1
  pivot, ans := (r+l)/2, []int{-1, -1}
  
  for l <= r {
    if nums[pivot] != target {
      if nums[pivot] > target {
        r = pivot-1
      } else {
        l = pivot+1
      }
      pivot = (r+l)/2
    } else {
      l, r = pivot, pivot
      for l >= 0 && nums[l] == target {
        l--
      }
      for r < len(nums) && nums[r] == target {
        r++
      }
      ans[0], ans[1] = l+1, r-1
      break
    }
  }
  return ans
}